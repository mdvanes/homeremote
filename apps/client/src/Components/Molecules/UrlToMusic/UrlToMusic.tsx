import React, { FC } from "react";
import {
    Button,
    Card,
    CardContent,
    TextField,
    LinearProgress,
} from "@mui/material";
import Alert from "@material-ui/lab/Alert";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../../Reducers";
import {
    UrlToMusicState,
    getInfo,
    setFormField,
    setFormFieldError,
    getMusic,
} from "./urlToMusicSlice";

const FILL_IN_THIS_FIELD = "Fill in this field";

const UrlToMusic: FC = () => {
    const dispatch = useDispatch();

    const {
        error: errorMessage,
        form,
        isLoading,
        result,
    } = useSelector<RootState, UrlToMusicState>(
        (state: RootState) => state.urlToMusic
    );

    const validateGetInfo = (): boolean => {
        if (form.url.value === "") {
            dispatch(setFormFieldError(["url", FILL_IN_THIS_FIELD]));
            return false;
        }
        return true;
    };

    const handleUrlChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(setFormField([event.target.name, event.target.value]));
    };

    const dispatchGetInfo = () => {
        const isValid = validateGetInfo();
        if (isValid) {
            dispatch(getInfo());
        }
    };

    const handleMusicDataChange = (
        event: React.ChangeEvent<HTMLInputElement>
    ) => {
        dispatch(setFormField([event.target.name, event.target.value]));
    };

    const validateGetMusic = (): boolean => {
        const invalidFieldNames = Object.entries(form)
            .filter(([, formField]) => formField.value === "")
            .map(([fieldName]) => fieldName);
        invalidFieldNames.forEach((fieldName) => {
            dispatch(setFormFieldError([fieldName, FILL_IN_THIS_FIELD]));
        });
        return invalidFieldNames.length === 0;
    };

    const dispatchGetMusic = () => {
        const isValid = validateGetMusic();
        if (isValid) {
            dispatch(getMusic());
        }
    };

    return (
        <Card className="card-dashboard-height">
            <CardContent>
                <form>
                    <TextField
                        data-testid="url"
                        label="URL"
                        name="url"
                        fullWidth={true}
                        value={form.url.value}
                        error={Boolean(form.url.error)}
                        helperText={form.url.error}
                        onChange={handleUrlChange}
                    />
                    <Button
                        data-testid="get-info"
                        color="primary"
                        onClick={dispatchGetInfo}
                    >
                        Get Info
                    </Button>
                </form>
                <form>
                    <TextField
                        data-testid="title"
                        label="Title"
                        name="title"
                        fullWidth={true}
                        value={form.title.value}
                        error={Boolean(form.title.error)}
                        helperText={form.title.error}
                        onChange={handleMusicDataChange}
                    />
                    <TextField
                        data-testid="artist"
                        label="Artist"
                        name="artist"
                        fullWidth={true}
                        value={form.artist.value}
                        error={Boolean(form.artist.error)}
                        helperText={form.artist.error}
                        onChange={handleMusicDataChange}
                    />
                    <TextField
                        label="Album"
                        name="album"
                        fullWidth={true}
                        value={form.album.value}
                        error={Boolean(form.album.error)}
                        helperText={form.album.error}
                        onChange={handleMusicDataChange}
                    />
                    <Button
                        data-testid="get-music"
                        color="primary"
                        onClick={dispatchGetMusic}
                    >
                        Get Music
                    </Button>
                </form>
                {errorMessage && <Alert severity="error">{errorMessage}</Alert>}
                {result && <Alert severity="success">Result in {result}</Alert>}
                {isLoading && <LinearProgress />}
            </CardContent>
        </Card>
    );
};

export default UrlToMusic;
