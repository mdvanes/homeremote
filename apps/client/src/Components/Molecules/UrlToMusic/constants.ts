export const FILL_IN_THIS_FIELD = "Fill in this field";
