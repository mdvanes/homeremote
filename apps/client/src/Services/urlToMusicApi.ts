import {
    UrlToMusicGetInfoArgs,
    UrlToMusicGetInfoResponse,
    UrlToMusicGetMusicArgs,
    UrlToMusicGetMusicProgressResponse,
    UrlToMusicGetMusicResponse,
} from "@homeremote/types";
import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/dist/query/react";
import { willAddCredentials } from "../devUtils";

export const urlToMusicApi = createApi({
    reducerPath: "urlToMusicApi",
    baseQuery: fetchBaseQuery({
        baseUrl: `${process.env.NX_BASE_URL}/api/urltomusic`,
        credentials: willAddCredentials(),
    }),
    endpoints: (builder) => ({
        getInfo: builder.query<
            UrlToMusicGetInfoResponse,
            UrlToMusicGetInfoArgs
        >({
            query: ({ url }) => `/getinfo/${url}`,
        }),
        getMusic: builder.query<
            UrlToMusicGetMusicResponse,
            UrlToMusicGetMusicArgs
        >({
            query: ({ url, artist, title, album }) =>
                `/getmusic/${url}?artist=${artist}&title=${title}&album=${album}`,
        }),
        getMusicProgress: builder.query<
            UrlToMusicGetMusicProgressResponse,
            UrlToMusicGetInfoArgs
        >({
            query: ({ url }) => `/getmusic/${url}/progress`,
        }),
    }),
});

export const { useGetInfoQuery, useGetMusicQuery, useGetMusicProgressQuery } =
    urlToMusicApi;
