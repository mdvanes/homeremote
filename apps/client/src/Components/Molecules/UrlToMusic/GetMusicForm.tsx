import {
    UrlToMusicGetInfoArgs,
    UrlToMusicGetMusicArgs,
} from "@homeremote/types";
import { Alert, Button, LinearProgress, TextField } from "@mui/material";
import { skipToken } from "@reduxjs/toolkit/dist/query";
import {
    useGetMusicProgressQuery,
    useGetMusicQuery,
} from "../../../Services/urlToMusicApi";
import { getErrorMessage } from "../../../Utils/getErrorMessage";
import React, { FC, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../../Reducers";
import { logError } from "../LogCard/logSlice";
import { FILL_IN_THIS_FIELD } from "./constants";
import {
    setFormField,
    setFormFieldError,
    UrlToMusicState,
} from "./urlToMusicSlice";

const UPDATE_INTERVAL_MS = 5 * 1000;

const GetMusicForm: FC = () => {
    const dispatch = useDispatch();
    const [query, setQuery] = useState<UrlToMusicGetMusicArgs>();

    const {
        form,
        isLoading: isProgressLoading,
        result,
    } = useSelector<RootState, UrlToMusicState>(
        (state: RootState) => state.urlToMusic
    );

    const args: UrlToMusicGetMusicArgs | typeof skipToken = query
        ? query
        : skipToken;
    const { refetch, isFetching, isLoading, error } = useGetMusicQuery(args);

    const progressArgs: UrlToMusicGetInfoArgs | typeof skipToken =
        isProgressLoading && form.url.value
            ? { url: encodeURIComponent(form.url.value) }
            : skipToken;
    useGetMusicProgressQuery(progressArgs, {
        pollingInterval: UPDATE_INTERVAL_MS,
    });

    const handleMusicDataChange = (
        event: React.ChangeEvent<HTMLInputElement>
    ) => {
        dispatch(setFormField([event.target.name, event.target.value]));
    };

    const validateGetMusic = (): boolean => {
        const invalidFieldNames = Object.entries(form)
            .filter(([, formField]) => formField.value === "")
            .map(([fieldName]) => fieldName);
        invalidFieldNames.forEach((fieldName) => {
            dispatch(setFormFieldError([fieldName, FILL_IN_THIS_FIELD]));
        });
        return invalidFieldNames.length === 0;
    };

    useEffect(() => {
        if (error) {
            dispatch(
                logError(`GetMusicForm failure: ${getErrorMessage(error)}`)
            );
        }
    }, [error, dispatch]);

    return (
        <form>
            <TextField
                data-testid="title"
                label="Title"
                name="title"
                fullWidth={true}
                value={form.title.value}
                error={Boolean(form.title.error)}
                helperText={form.title.error}
                variant="standard"
                onChange={handleMusicDataChange}
            />
            <TextField
                data-testid="artist"
                label="Artist"
                name="artist"
                fullWidth={true}
                value={form.artist.value}
                error={Boolean(form.artist.error)}
                helperText={form.artist.error}
                variant="standard"
                onChange={handleMusicDataChange}
            />
            <TextField
                label="Album"
                name="album"
                fullWidth={true}
                value={form.album.value}
                error={Boolean(form.album.error)}
                helperText={form.album.error}
                variant="standard"
                onChange={handleMusicDataChange}
            />
            <Button
                data-testid="get-music"
                color="primary"
                onClick={() => {
                    const isValid = validateGetMusic();
                    if (isValid) {
                        setQuery({
                            url: encodeURIComponent(form.url.value),
                            title: encodeURIComponent(form.title.value),
                            artist: encodeURIComponent(form.artist.value),
                            album: encodeURIComponent(form.album.value),
                        });
                        // To force ignoring cached value
                        refetch();
                    }
                }}
            >
                Get Music
            </Button>
            {!isLoading && !isFetching && (
                <>
                    {error && (
                        <Alert severity="error">{getErrorMessage(error)}</Alert>
                    )}
                </>
            )}
            {(isLoading || isFetching || isProgressLoading) && (
                <LinearProgress />
            )}
            {result && <Alert severity="success">Result in {result}</Alert>}
        </form>
    );
};

export default GetMusicForm;
