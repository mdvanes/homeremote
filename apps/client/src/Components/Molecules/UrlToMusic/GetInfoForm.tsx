import { UrlToMusicGetInfoArgs } from "@homeremote/types";
import { Alert, Button, CircularProgress, TextField } from "@mui/material";
import { skipToken } from "@reduxjs/toolkit/dist/query";
import React, { FC, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../../Reducers";
import { useGetInfoQuery } from "../../../Services/urlToMusicApi";
import { getErrorMessage } from "../../../Utils/getErrorMessage";
import { logError } from "../LogCard/logSlice";
import { FILL_IN_THIS_FIELD } from "./constants";
import {
    reset,
    setFormField,
    setFormFieldError,
    UrlToMusicState,
} from "./urlToMusicSlice";

const GetInfoForm: FC = () => {
    const dispatch = useDispatch();
    const [query, setQuery] = useState<string | undefined>();

    const { form } = useSelector<RootState, UrlToMusicState>(
        (state: RootState) => state.urlToMusic
    );
    const args: UrlToMusicGetInfoArgs | typeof skipToken = query
        ? { url: query }
        : skipToken;

    const { isFetching, isLoading, error, refetch } = useGetInfoQuery(args);

    const validateGetInfo = (): boolean => {
        if (form.url.value === "") {
            dispatch(setFormFieldError(["url", FILL_IN_THIS_FIELD]));
            return false;
        }
        return true;
    };

    const handleUrlChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(setFormField([event.target.name, event.target.value]));
    };

    useEffect(() => {
        if (error) {
            dispatch(
                logError(`GetInfoForm failure: ${getErrorMessage(error)}`)
            );
        }
    }, [error, dispatch]);

    return (
        <>
            <form>
                <TextField
                    data-testid="url"
                    label="URL"
                    name="url"
                    fullWidth={true}
                    value={form.url.value}
                    error={Boolean(form.url.error)}
                    helperText={form.url.error}
                    variant="standard"
                    onChange={handleUrlChange}
                />
                <Button
                    data-testid="get-info"
                    color="primary"
                    onClick={() => {
                        const isValid = validateGetInfo();
                        if (isValid) {
                            dispatch(reset());
                            setQuery(encodeURIComponent(form.url.value));
                            // Always call refetch to ignore cache and overwrite the music info
                            refetch();
                        }
                    }}
                >
                    Get Info
                </Button>
            </form>
            {error && <Alert severity="error">{getErrorMessage(error)}</Alert>}
            {(isFetching || isLoading) && <CircularProgress />}
            {/* TODO show a sensible preview */}
            {/* {data?.streamUrl &&
                data.streamUrl.map((s) => (
                    <video width={400} controls src={s} />
                ))} */}
        </>
    );
};

export default GetInfoForm;
