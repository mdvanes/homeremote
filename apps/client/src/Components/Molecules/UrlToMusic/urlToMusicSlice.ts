import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import ApiBaseState from "../../../Reducers/state.types";
import { urlToMusicApi } from "../../../Services/urlToMusicApi";

interface FormField {
    value: string;
    error: string | false;
}

export interface UrlToMusicState extends ApiBaseState {
    form: Record<string, FormField>;
    showGetMusicForm: boolean;
    result: string | false;
}

const year = new Date().getFullYear();

export const initialState: UrlToMusicState = {
    isLoading: false,
    error: false,
    showGetMusicForm: false,
    form: {
        url: {
            value: "",
            error: false,
        },
        title: {
            value: "",
            error: false,
        },
        artist: {
            value: "",
            error: false,
        },
        album: {
            value: `Songs from ${year}`,
            error: false,
        },
    },
    result: false,
};

type NameAndValue = [string, string];

const urlToMusicSlice = createSlice({
    name: "urlToMusic",
    initialState,
    reducers: {
        reset: (draft) => {
            const url = draft.form.url.value;
            return {
                ...initialState,
                form: {
                    ...initialState.form,
                    url: {
                        ...initialState.form.url,
                        value: url,
                    },
                },
            };
        },
        resetAll: () => {
            return initialState;
        },
        resetProgressLoading: (draft) => {
            draft.isLoading = false;
        },
        setFormField: (
            draft,
            { payload }: PayloadAction<NameAndValue>
        ): void => {
            const [name, value] = payload;
            draft.form[name].value = value;
            draft.form[name].error = "";
        },
        setFormFieldError: (
            draft,
            { payload }: PayloadAction<NameAndValue>
        ): void => {
            const [name, error] = payload;
            draft.form[name].error = error;
        },
    },
    extraReducers: (builder) => {
        builder.addMatcher(
            urlToMusicApi.endpoints.getInfo.matchFulfilled,
            (draft, { payload }): void => {
                draft.isLoading = initialState.isLoading;
                draft.form.title.value = payload.title;
                draft.form.title.error = false;
                draft.form.artist.value = payload.artist;
                draft.form.artist.error = false;
                draft.showGetMusicForm = true;
            }
        );
        builder.addMatcher(
            urlToMusicApi.endpoints.getMusic.matchFulfilled,
            (draft, { payload }): void => {
                draft.isLoading = true;
            }
        );
        builder.addMatcher(
            urlToMusicApi.endpoints.getMusicProgress.matchFulfilled,
            (draft, { payload }): void => {
                if (payload.state === "finished") {
                    draft.isLoading = initialState.isLoading;
                    draft.result = payload.path;
                }
            }
        );
    },
});

export const {
    setFormField,
    setFormFieldError,
    reset,
    resetAll,
    resetProgressLoading,
} = urlToMusicSlice.actions;

export default urlToMusicSlice.reducer;
