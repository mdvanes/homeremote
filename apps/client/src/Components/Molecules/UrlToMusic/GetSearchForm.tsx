import { UrlToMusicGetSearchArgs } from "@homeremote/types";
import {
    Close as CloseIcon,
    FormatQuote as FormatQuoteIcon,
    MusicNote as MusicNoteIcon,
    Search as SearchIcon,
} from "@mui/icons-material";
import {
    Alert,
    Button,
    CardActions,
    CircularProgress,
    InputAdornment,
    List,
    ListItemButton,
    ListItemText,
    TextField,
} from "@mui/material";
import { skipToken } from "@reduxjs/toolkit/dist/query";
import { FC, useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import {
    urlToMusicApi,
    useGetSearchQuery,
} from "../../../Services/urlToMusicApi";
import { getErrorMessage } from "../../../Utils/getErrorMessage";
import { logError } from "../LogCard/logSlice";
import { resetAll, setFormField } from "./urlToMusicSlice";

const GetSearchForm: FC = () => {
    const dispatch = useDispatch();
    const [terms, setTerms] = useState<string>();
    const [query, setQuery] = useState<string>();

    const handleTermsChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setTerms(event.target.value);
    };

    const handlePrefill = async () => {
        const nowplaying: { artist: string; title: string } = await fetch(
            `${process.env.NX_BASE_URL}/api/nowplaying/radio2`
        ).then((data) => data.json());
        const newTerm = `${nowplaying.artist} ${nowplaying.title}`;
        setTerms(newTerm);
    };

    const args: UrlToMusicGetSearchArgs | typeof skipToken = query
        ? { terms: query }
        : skipToken;
    const {
        data: result,
        isLoading,
        isFetching,
        error,
    } = useGetSearchQuery(args);

    const resultElems = result?.searchResults ? (
        <List dense>
            {result.searchResults.map((item, index) => (
                <ListItemButton
                    key={index}
                    sx={{ gap: 2 }}
                    onClick={() => {
                        dispatch(
                            setFormField([
                                "url",
                                `https://youtube.com/watch?v=${item.id}`,
                            ])
                        );
                    }}
                >
                    <ListItemText primary={item.title} />
                </ListItemButton>
            ))}
        </List>
    ) : null;

    useEffect(() => {
        if (error) {
            dispatch(
                logError(
                    `GetSearchForm failure: ${getErrorMessage(
                        error
                    ).toString()}`
                )
            );
        }
    }, [dispatch, error]);

    if (error) {
        return <Alert severity="error">{getErrorMessage(error)}</Alert>;
    }

    return (
        <form
            onSubmit={(ev) => {
                ev.preventDefault();
                setQuery(terms);
            }}
        >
            <TextField
                data-testid="terms"
                label="Search terms"
                name="terms"
                fullWidth={true}
                // preventing undefined will prevent overlapping the label with the input
                value={terms ?? ""}
                variant="standard"
                onChange={handleTermsChange}
                InputProps={{
                    endAdornment: (
                        <InputAdornment position="end">
                            <MusicNoteIcon />
                        </InputAdornment>
                    ),
                }}
            />
            {isLoading || isFetching ? (
                <CircularProgress sx={{ marginTop: 2 }} />
            ) : (
                resultElems
            )}
            <CardActions>
                <Button
                    startIcon={<SearchIcon />}
                    color="primary"
                    onClick={() => {
                        setQuery(terms);
                    }}
                >
                    search
                </Button>
                <Button
                    startIcon={<FormatQuoteIcon />}
                    color="primary"
                    onClick={handlePrefill}
                >
                    prefill
                </Button>
                <Button
                    startIcon={<CloseIcon />}
                    color="warning"
                    onClick={() => {
                        setTerms("");
                        setQuery("");
                        dispatch(resetAll());
                        urlToMusicApi.util.resetApiState();
                    }}
                >
                    reset
                </Button>
            </CardActions>
        </form>
    );
};

export default GetSearchForm;
