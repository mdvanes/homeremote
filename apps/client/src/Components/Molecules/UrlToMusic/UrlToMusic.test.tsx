import { UrlToMusicGetInfoResponse } from "@homeremote/types";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import fetchMock, { enableFetchMocks } from "jest-fetch-mock";
import { FC } from "react";
import { urlToMusicApi } from "../../../Services/urlToMusicApi";
import {
    createGetCalledUrl,
    MockStoreProvider,
    MockStoreProviderApi,
} from "../../../testHelpers";
import loglinesReducer from "../../Molecules/LogCard/logSlice";
import UrlToMusic from "./UrlToMusic";
import urlToMusicReducer from "./urlToMusicSlice";

enableFetchMocks();

const getCalledUrl = (callNr: number) => createGetCalledUrl(fetchMock)(callNr);

const urlToMusicSliceFakeApi: MockStoreProviderApi = {
    reducerPath: "urlToMusic",
    reducer: urlToMusicReducer,
};

const loglinesFakeApi: MockStoreProviderApi = {
    reducerPath: "loglines",
    reducer: loglinesReducer,
};

const Wrapper: FC = ({ children }) => {
    return (
        <MockStoreProvider
            apis={[urlToMusicApi, urlToMusicSliceFakeApi, loglinesFakeApi]}
        >
            {children}
        </MockStoreProvider>
    );
};

describe("UrlToMusic", () => {
    beforeEach(() => {
        fetchMock.resetMocks();
    });

    it("shows the GetInfo form", () => {
        const { asFragment } = render(<UrlToMusic />, { wrapper: Wrapper });

        expect(asFragment()).toMatchSnapshot();
    });

    const expectSetUrl = async () => {
        const rendered = render(<UrlToMusic />, { wrapper: Wrapper });

        const urlInput = screen.getByRole("textbox", { name: "URL" });
        expect(
            screen.queryByRole("textbox", {
                name: "Title",
            })
        ).not.toBeInTheDocument();

        if (!urlInput) {
            throw Error("missing input");
        }

        fireEvent.change(urlInput, { target: { value: "Some URL" } });
        // userEvent.type(urlInput, "Some URL");
        // userEvent.tab();

        expect(fetchMock).not.toBeCalled();
        const getInfoButton = screen.getByRole("button", { name: "Get Info" });
        return { urlInput, getInfoButton, rendered };
    };

    const mockGetInfoResponse: UrlToMusicGetInfoResponse = {
        title: "Some Title",
        artist: "Some Artist",
        streamUrl: ["a"],
        versionInfo: "1",
    };

    it("can submit the getinfo form", async () => {
        const { urlInput, getInfoButton, rendered } = await expectSetUrl();

        fetchMock.mockResponse(JSON.stringify(mockGetInfoResponse));
        await userEvent.click(getInfoButton);

        const titleInput = await screen.findByRole("textbox", {
            name: "Title",
        });

        expect(urlInput).toHaveValue("Some URL");
        expect(rendered.asFragment()).toMatchSnapshot(
            "shows the GetMusic form"
        );

        expect(titleInput).toHaveValue("Some Title");
        expect(screen.getByRole("textbox", { name: "Artist" })).toHaveValue(
            "Some Artist"
        );
        expect(fetchMock).toBeCalledTimes(1);
        expect(getCalledUrl(0)).toBe(
            "http://localhost/api/urltomusic/getinfo/Some%20URL"
        );
    });

    it("can handle getinfo error", async () => {
        const { getInfoButton } = await expectSetUrl();

        fetchMock.mockReject(Error("some getinfo error"));
        await userEvent.click(getInfoButton);

        expect(
            screen.getByText("[FETCH_ERROR] Error: some getinfo error")
        ).toBeVisible();
    });

    it("can submit the getmusic form", async () => {
        const { getInfoButton } = await expectSetUrl();

        fetchMock.mockResponse(JSON.stringify(mockGetInfoResponse));
        await userEvent.click(getInfoButton);

        const titleInput = await screen.findByRole("textbox", {
            name: "Title",
        });

        if (titleInput) {
            fireEvent.change(titleInput, {
                target: { value: "Some Other Title" },
            });
        }

        const getMusicButton = screen.getByRole("button", {
            name: "Get Music",
        });

        fetchMock.mockReset();
        // const mockGetMusicResponse: UrlToMusicGetMusicResponse = {};
        // fetchMock.mockResponses(JSON.stringify(mockGetInfoResponse));
        await userEvent.click(getMusicButton);

        expect(fetchMock).toBeCalledTimes(1);
        expect(getCalledUrl(0)).toBe(
            "http://localhost/api/urltomusic/getmusic/Some%20URL?artist=Some%20Artist&title=Some%20Other%20Title&album=Songs%20from%202022"
        );
    });
});
