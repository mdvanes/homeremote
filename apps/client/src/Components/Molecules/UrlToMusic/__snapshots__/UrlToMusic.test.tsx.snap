// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlToMusic can submit the getinfo form: shows the GetMusic form 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card-dashboard-height css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <form>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="url"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-2"
            id="mui-2-label"
          >
            URL
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-2"
              name="url"
              type="text"
              value="Some URL"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          data-testid="get-info"
          tabindex="0"
          type="button"
        >
          Get Info
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          >
            <span
              class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
              style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
            >
              <span
                class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
              />
            </span>
          </span>
        </button>
      </form>
      <form>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="title"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-3"
            id="mui-3-label"
          >
            Title
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-3"
              name="title"
              type="text"
              value="Some Title"
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="artist"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-4"
            id="mui-4-label"
          >
            Artist
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-4"
              name="artist"
              type="text"
              value="Some Artist"
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-5"
            id="mui-5-label"
          >
            Album
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-5"
              name="album"
              type="text"
              value="Songs from 2022"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          data-testid="get-music"
          tabindex="0"
          type="button"
        >
          Get Music
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`UrlToMusic shows the GetInfo form 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card-dashboard-height css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <form>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="url"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-1"
            id="mui-1-label"
          >
            URL
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-1"
              name="url"
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          data-testid="get-info"
          tabindex="0"
          type="button"
        >
          Get Info
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
